leasing/report/invoice/invoicing_review.py

```diff
@@ -423,6 +440,124 @@ def get_incorrect_management_shares_data(self, service_unit_ids, cursor):
 
         return data
 
+    def get_gaps_in_billing_periods_data(self, service_unit_ids, cursor):
+
+        query = """
+            SELECT li.identifier AS "lease_identifier",
+                l.id AS "lease_id",
+                l.start_date,
+                l.end_date,
+                r.id AS "rent_id",
+                r.start_date AS "rent_start_date",
+                r.end_date AS "rent_end_date",
+                i.id AS "invoice_id",
+                i.billing_period_start_date,
+                i.billing_period_end_date
+            FROM leasing_lease l
+            INNER JOIN leasing_leaseidentifier li
+                ON l.identifier_id = li.id
+            INNER JOIN leasing_rent r
+                ON l.id = r.lease_id
+            INNER JOIN leasing_invoice i
+                ON l.id = i.lease_id
+            INNER JOIN leasing_invoicerow ir
+                ON i.id = ir.invoice_id
+            WHERE (l.end_date IS NULL OR l.end_date >= %(today)s)
+                AND (l.start_date IS NOT NULL AND l.start_date <= %(today)s)
+                AND l.service_unit_id = ANY(%(service_units)s)
+                AND l.deleted IS NULL
+                AND l.state NOT IN ('reservation', 'power_of_attorney')
+                AND r.deleted IS NULL
+                AND r."type" != 'free'
+                AND (r.start_date IS NULL OR r.start_date <= %(today)s)
+                AND (r.end_date IS NULL OR r.end_date >= %(today)s)
+                AND i.deleted IS NULL
+                AND i.deleted_by_cascade IS false
+                AND ir.receivable_type_id != ANY(%(excluded_receivable_type_ids)s)
+            ORDER BY li.identifier ASC;
+        """
+
+        today = datetime.date.today()
+
+        excluded_receivable_type_ids = [
+            id
+            for id in ReceivableType.objects.filter(
+                name__in=EXCLUDED_RECEIVABLE_TYPE_NAMES
+            ).values_list("id", flat=True)
+        ]
+
+        cursor.execute(
+            query,
+            {
+                "service_units": service_unit_ids,
+                "today": today,
+                # convert ids to string for the query
+                "excluded_receivable_type_ids": [
+                    id for id in excluded_receivable_type_ids
+                ],
+            },
+        )
+
+        rows = dictfetchall(cursor)
+        row_indexes = range(len(rows))
+
+        data = []
+
+        active_rent_period_start = None
+        active_rent_period_end = None
+        active_rent_period_days = 0
+        invoiced_period_days = 0
+        lease_has_gaps = False
```

Comment:
I find this a bit odd to be defined outside of the for-loop, why is it outside of the for-loop?
If I am reading this correctly sometimes it gets set as True for the next iteration of the loop, for the next lease. Is that intentional behavior (for example line 528)?

Reply:
Yes, it is. There are several rows for each lease, so it cannot be re-assigned as False at the beginning of each iteration. If a gap is found on one of the rows, it should be assigned as True, and kept so until the last row for the current lease is reached, so that it could be added to the final list of results.

Reply:
https://docs.python.org/3/library/itertools.html#itertools.groupby

Reply:
I didn't create an inner loop, but I moved the declaration of these variables inside the loop to be defined at index 0.